File: eldoria/world/perk_effects.py
"""Applies a chosen Perk's permanent effect.

Most perks bake a flat stat bump in immediately. SECOND_WIND, MASTER_TRADER,
and CRITICAL_FOCUS don't touch a stat here -- they're read live at the point
of use instead (a near-death save, a shop discount, a combat crit-threshold
check via PlayerCharacter.perk_rank).
"""
from __future__ import annotations

from dataclasses import replace

from eldoria.models import Perk, PlayerCharacter, SkillType
from eldoria.world.skill_progression import MAX_SKILL_LEVEL


def apply_perk(player: PlayerCharacter, perk: Perk) -> PlayerCharacter:
    if player.pending_perk_choices <= 0:
        raise ValueError("No perk choice banked")

    new_perks = dict(player.perks)
    new_perks[perk] = new_perks.get(perk, 0) + 1
    with_perk = replace(player, perks=new_perks, pending_perk_choices=player.pending_perk_choices - 1)

    if perk == Perk.POWER_ATTACK:
        return replace(with_perk, attack_bonus=with_perk.attack_bonus + 2)
    if perk == Perk.IRON_SKIN:
        return replace(with_perk, armor_class=with_perk.armor_class + 1)
    if perk == Perk.QUICK_REFLEXES:
        return replace(with_perk, speed=with_perk.speed + 3)
    if perk == Perk.ARCANE_RESERVE:
        return replace(with_perk, willpower=with_perk.willpower + 2)
    if perk == Perk.SILENT_STEP:
        sneak = with_perk.skills.get(SkillType.SNEAK)
        if sneak is None:
            return with_perk
        new_skills = dict(with_perk.skills)
        new_skills[SkillType.SNEAK] = replace(sneak, level=min(MAX_SKILL_LEVEL, sneak.level + 15))
        return replace(with_perk, skills=new_skills)
    if perk == Perk.TOUGHNESS:
        return replace(with_perk, max_health=with_perk.max_health + 15, current_health=with_perk.current_health + 15)
    # SECOND_WIND, MASTER_TRADER, CRITICAL_FOCUS: behavioral flags only, checked live elsewhere.
    return with_perk

File: core/src/commonMain/kotlin/eldoria/core/world/PerkEffects.kt
package eldoria.core.world

import eldoria.core.model.Perk
import eldoria.core.model.PlayerCharacter
import eldoria.core.model.SkillType

/**
 * Applies a chosen Perk's permanent effect. Most perks bake a flat stat
 * bump in immediately (same "recompute once, store the result" style as
 * LevelProgression) -- and since `perks` is now a stack count (see
 * PlayerCharacter.perks' doc), calling this again for the same perk just
 * bakes the bump in again, which is already correct with no extra code:
 * every one of these is a delta add, not a "set to X". SECOND_WIND,
 * MASTER_TRADER, and CRITICAL_FOCUS don't touch a stat here -- they're
 * read live at the point of use instead (a near-death save, a shop
 * discount, a combat crit-threshold check via PlayerCharacter.perkRank).
 */
object PerkEffects {
    fun applyPerk(player: PlayerCharacter, perk: Perk): PlayerCharacter {
        require(player.pendingPerkChoices > 0) { "No perk choice banked" }
        val withPerk = player.copy(
            perks = player.perks + (perk to (player.perks[perk] ?: 0) + 1),
            pendingPerkChoices = player.pendingPerkChoices - 1,
        )
        return when (perk) {
            Perk.POWER_ATTACK -> withPerk.copy(attackBonus = withPerk.attackBonus + 2)
            Perk.IRON_SKIN -> withPerk.copy(armorClass = withPerk.armorClass + 1)
            Perk.QUICK_REFLEXES -> withPerk.copy(speed = withPerk.speed + 3)
            Perk.ARCANE_RESERVE -> withPerk.copy(willpower = withPerk.willpower + 2)
            Perk.SILENT_STEP -> {
                val sneak = withPerk.skills[SkillType.SNEAK]
                if (sneak == null) withPerk else withPerk.copy(
                    skills = withPerk.skills + (SkillType.SNEAK to sneak.copy(level = (sneak.level + 15).coerceAtMost(SkillProgression.MAX_SKILL_LEVEL))),
                )
            }
            Perk.TOUGHNESS -> withPerk.copy(maxHealth = withPerk.maxHealth + 15, currentHealth = withPerk.currentHealth + 15)
            Perk.SECOND_WIND -> withPerk // behavioral flag only, checked via `player.perkRank(SECOND_WIND) > 0`
            Perk.MASTER_TRADER -> withPerk // behavioral flag only, checked by ShopGenerator/shop commands
            Perk.CRITICAL_FOCUS -> withPerk // read live: Game.kt's attack() lowers the crit threshold by player.perkRank(CRITICAL_FOCUS)
        }
    }
}
